[Merged by Bors] - feat: add some name validation for authors

adomani opened the pull request:
The check consists of verifying that each file has at least one Copyright holder and one Author.

---

- [x] depends on: ##31252
- [x] depends on: ##31253
- [x] depends on: ##31255

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
feat: add some name validation for authors

adomani committed:
restore test

adomani committed:
fixes and exceptions

adomani committed:
remove test

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into lint_more_names_in_header_linter

grunweg reviewed (commented):
Good idea, and a very tricky topic. I left some comments - mostly about error messaging. The current check is broad, but that seems fine to me (and perhaps better than trying otherwise).

adomani committed:
lint less: only check that there is a non-empty string of names

adomani commented on MathlibTest/Header.lean:
```diff
@@ -320,3 +337,30 @@ Authors: Name LastName
+---
+info: Text: `Authors`
+Range: (140, 147)
+Message: 'We are aware that standardizing names is hard (https:||www.kalzumeus.com|2010|06|17|falsehoods-programmers-believe-about-names|). We are happy to add your name as an exception to the linter, though!'
```
The link disappeared, since we are no longer actually linting on the name, but let me answer this nonetheless. The `#check_copyright` replaces `/` with `|` since otherwise, with the `#guard_msgs`, a substring `-/` would actually close the guarded message and would create havoc in the tests.

The conclusion is that when you see `|` it may come from `/` in the source and it does in this case.

grunweg replied:
:bulb: Thanks for reminding me!

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into lint_more_names_in_header_linter

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into lint_more_names_in_header_linter

grunweg committed:
Three more tests

grunweg committed:
Merge branch 'master' into lint_more_names_in_header_linter

grunweg committed:
Merge branch 'master' into lint_more_names_in_header_linter

grunweg commented:
The copyright violations have been fixed now, so hopefully this PR is good to go :tada:

grunweg commented:
bors d+

grunweg committed:
chore(GroupTheory/PGroup): fix copyright

Add Chris Hughes as copyright holder: he as the initial author of most
lemmas in this file.
The current module doc-string goes back to the port, and back to mathlib#8595
which split this out (and put no copyright holder). The particular lemmas were
created in mathlib#591 and other PRs by Chris.

grunweg committed:
chore: fix formatting of one copyright header

grunweg committed:
chore(Algebra/PresentedMonoid): fix copyright header

This file was added in #13152: it appears that the author simply forgot to
add themselves as the copyright holder.

grunweg committed:
chore(Data/SetLike/Fintype): fix copyright header

This instance was added in mathlib##8593; it appears the main author
simply forgot to add themselves as copyright holder.

grunweg committed:
chore(RingTheory/TwoSidedIdeal/Instances): fix copyright header

This file was added in #15986, and it appears the main author simply
forgot to add themselves as the copyright holder.

grunweg committed:
chore(Tactic/Group): fix copyright header

Patrick Massot write the tactic in Lean 3 (in mathlib#3062); Thomas
Browning ported it to Lean 4. The copyright holder was simply omitted
in the initial PR, so this appears a mere oversight.

grunweg committed:
chore(Tactic/CategoryTheory/Coherence): fix copyright headers

Both files were originally one file, the Lean 4 port of the coherence tactic.
That one was initially written by Kim Morrison in mathlib#13125.
It appears that omitting the copyright holder was a mere oversight.

grunweg committed:
Merge branch 'fix-copyright-clearcut' into lint_more_names_in_header_linter

grunweg committed:
Fix test: thanks, Markus Himmel, for the help!

grunweg commented:
Green!
bors merge
